### PR TITLE
feat: support self-signed certificates

### LINKED ISSUE #563: Support for self-signed certs on self-hosted gitlab instance
Documentation should indicate how to configure the certificate for self-signed certs on gitlab server.

Gitlab supports the following approaches:
https://docs.gitlab.com/runner/configuration/tls-self-signed.html#supported-options-for-self-signed-certificates-targeting-the-gitlab-server

Currently, `tls-ca-file` setting is not available in the module. Perhaps it can be added.
Documentation should be updated either way, and show a working example to make the module easier to use.

### PR DESCRIPTION
**Pull request by baolsen**
## Description

This change simply allows the Gitlab runner to register and run with a self-signed gitlab instance. The runner helper image can access such an instance.

Closes #563

## Migrations required

NO

## Verification

I've verified that the iniital change works in my own environment where I have self-signed certificates in use.
I have not verified that the changes after enabling pre-commit and tf docs etc work, but they should...

### CODE REVIEW

**Comment by baolsen** [comment]
Just a note, I could only get this to work with the `docker` executor.

With `docker+machine` I tried but could not find a way to install the certs on the child runners, I think it is not natively supported and no longer maintained either.

Perhaps if someone else in future wants this functionality they can provide a PR for it.

File: examples/runner-certificates/README.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+----
+### Resources
+
+| Name | Type |
+|------|------|
+| [aws_availability_zones.available](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/availability_zones) | data source |
+
+----
+<!-- END_TF_DOCS -->
```
Comment (kayman-mk) [inline examples/runner-certificates/README.md:216]: suggestion: remove everything between the `TF_DOCS` tags. It's automatically generated as soon as the branch is merged into `develop`.

File: examples/runner-certificates/README.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+**Option 2:** Adding a section to each pipeline using `before_script`:
+```yaml
+default:
+  before_script:
+    - apt-get install -y ca-certificates
+    - cp /etc/gitlab-runner/certs/* /usr/local/share/ca-certificates/
+    - update-ca-certificates
+```
+
```
Comment (kayman-mk) [inline examples/runner-certificates/README.md:87]: thought: mention a third option which uses the script from option 2 but put it into the `runners_pre_build_script`

File: examples/runner-certificates/README.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
+# Example - Spot Runner - Private subnet
+
+In this scenario the runner agent is running on a single EC2 node.
+
+The example is intended to show how the runner can be configured for self-hosted Gitlab environments with certificates signed by a custom CA.
+
```
Comment (kayman-mk) [inline examples/runner-certificates/README.md:6]: suggestion: as you wrote in the PR, this is not usable for the `docker+machine` executor. Could you please add a note?

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -1,12 +1,45 @@
+# https://pre-commit.com/
```
Comment (kayman-mk) [inline .pre-commit-config.yaml:1]: suggestion: please move the file into another PR as it is not related to the certificate problem tackled here.

The linters are really useful and tackled in #580.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -7,8 +7,23 @@ on:
...
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+      - uses: hashicorp/setup-terraform@v2
+      - name: Run
+        run: |
+          python -m venv .venv
+          source .venv/bin/activate
+          pip install pre-commit checkov
+          pre-commit install
+          pre-commit run --all-files --show-diff-on-failure
```
Comment (kayman-mk) [inline .github/workflows/ci.yml:23]: thought: should be removed as it is tackled in #580 and not related to the certificate problem.

File: locals.tf [hunk 6]
```diff
@@ -1,4 +1,45 @@
 locals {
+  # Manage certificates
+  pre_install_certificate_1 = (
```
Comment (kayman-mk) [inline locals.tf:3]: issue: I think something like `pre_install_gitlab_certificate` makes it clearer.

File: locals.tf [hunk 7]
```diff
@@ -1,4 +1,45 @@
...
+    length(var.runners_gitlab_certificate) > 0
+    ? <<-EOT
+      mkdir -p /etc/gitlab-runner/certs/
+      cat <<- EOF > /etc/gitlab-runner/certs/gitlab.crt
+      ${var.runners_gitlab_certificate}
+      EOF
+    EOT
+    : ""
+  )
+  pre_install_certificate_2 = (
```
Comment (kayman-mk) [inline locals.tf:13]: issue: I think something like `pre_install_ca_certificates` makes it clearer.

File: locals.tf [hunk 8]
```diff
@@ -1,4 +1,45 @@
...
+    ]
+  )
+  pre_install_certificates_end = <<-EOT
+    chmod 600 /etc/gitlab-runner/certs/*.crt
+    chmod -R a+r /etc/gitlab-runner
+    cp /etc/gitlab-runner/certs/*.crt /etc/pki/ca-trust/source/anchors
+    update-ca-trust extract
+  EOT
+  pre_install_certificates = (
+    length(local.pre_install_certificates_all) > 5 # Newlines
```
Comment (kayman-mk) [inline locals.tf:35]: suggestion: we could quickly run into problems here with the hardcoded `5`. What about something like `length(var.runners_gitlab_certificate) +length(var.runners_ca_certificate) > 0`? Then you could also get rid of the `pre_install_certificates_all` variable.

File: examples/runner-certificates/README.md [hunk 9]
```diff
@@ -0,0 +1,216 @@
+# Example - Spot Runner - Private subnet
```
Comment (kayman-mk) [inline examples/runner-certificates/README.md:1]: praise: appreciated! Having an example makes life easier. Thanks for that.

Comment (baolsen) [reply examples/runner-certificates/README.md:1]: @kayman-mk thanks for the great feedback :) 

I agree and I've updated for all suggestions, please check again.

**Comment by kayman-mk** [comment]
I will do a quick check in my environment. Code looks good.

**Comment by kayman-mk** [comment]
Verified that my runners in a complex scenario are still running (without using the new feature): Success

**Review by kayman-mk — APPROVED** [review]

**Comment by npalm** [comment]
Tested the PR on the default branch, docker machines are not starting healthy.

**Review by npalm — CHANGES_REQUESTED** [review]
Tested the PR on the default branch, docker machines are not starting healthy. Build are not starting.

```
NAME                                        ACTIVE   DRIVER      STATE     URL                    SWARM   DOCKER    ERRORS
runner-gksonazu-npalm-1670767825-0f2da2d9   -        amazonec2   Running   tcp://10.0.1.93:2376           Unknown   Unable to query dockerversion: Cannot connect to the docker engine endpoint
```

So seems the change in the start scripts breaks the runners.

**Comment by npalm** [comment]
> Tested the PR on the default branch, docker machines are not starting healthy. Build are not starting.
> 
> ```
> NAME                                        ACTIVE   DRIVER      STATE     URL                    SWARM   DOCKER    ERRORS
> runner-gksonazu-npalm-1670767825-0f2da2d9   -        amazonec2   Running   tcp://10.0.1.93:2376           Unknown   Unable to query dockerversion: Cannot connect to the docker engine endpoint
> ```
> 
> So seems the change in the start scripts breaks the runners.

Another run, now it seems working. Would like to check the PR in a bit more detail

**Comment by jpooler** [comment]
@npalm  any chance this one can get merged in? being able to customize certs makes dind possible, among other things.

**Comment by kayman-mk** [comment]
@npalm Same here. Merged the current `main` in, but the pipelines are no longer starting. Seems to be a problem with the cloud init script.

**Comment by kayman-mk** [comment]
@npalm Could you please recheck? Looks good here. Tested the "no own certificate" path only. So nobody should be affected by this change.

**Review by kayman-mk — APPROVED** [review]

File: template/gitlab-runner.tpl [hunk 10]
```diff
@@ -20,22 +20,29 @@ EOF
...
 
+${pre_install_certificates}
+
 # fetch Runner token from SSM and validate it
 token=$(aws ssm get-parameters --names "${secure_parameter_store_runner_token_key}" --with-decryption --region "${secure_parameter_store_region}" | jq -r ".Parameters | .[0] | .Value")
 
 valid_token=true
 if [[ "$token" != "null" ]]
 then
+<<<<<<< HEAD
```
Comment (npalm) [inline template/gitlab-runner.tpl:31]: The PR contains several incorrect merges, please can you fix them.

**Review by npalm — CHANGES_REQUESTED** [review]
@baolsen sorry for the late response. But it seems the PR is not merged correctly. I made on one place a comment, but it seems there are multiple places where the reabase went wrong. Please can you fix the rebase?

File: template/gitlab-runner.tpl (hunk 10 above)
Comment (kayman-mk) [reply template/gitlab-runner.tpl:31]: Ups, I guess it was me. Have to activate the `pre-commit` immediately ...

Comment (kayman-mk) [reply template/gitlab-runner.tpl]: One incorrect merge found and fixed.

**Comment by npalm** [comment]
@kayman-mk I also checked the default example. For me the PR is OK. But the only thing that is strange is. When I do a local rebase with main, I got merge conflicts. But form reason they are not shown here.

**Review by npalm — APPROVED** [review]

**Comment by npalm** [comment]
@baolsen thank you for you contribution!
